File: app/src/main/java/org/jellyfin/androidtv/ui/presentation/MyDetailsOverviewRowPresenter.java
package org.jellyfin.androidtv.ui.presentation;

import android.text.TextUtils;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;
import android.widget.ProgressBar;
import android.widget.RelativeLayout;
import android.widget.TextView;

import androidx.leanback.widget.RowPresenter;

import org.jellyfin.androidtv.data.model.InfoItem;
import org.jellyfin.androidtv.databinding.ViewRowDetailsBinding;
import org.jellyfin.androidtv.ui.AsyncImageView;
import org.jellyfin.androidtv.ui.DetailRowView;
import org.jellyfin.androidtv.ui.TextUnderButton;
import org.jellyfin.androidtv.ui.itemdetail.MyDetailsOverviewRow;
import org.jellyfin.androidtv.util.InfoLayoutHelper;
import org.jellyfin.androidtv.util.MarkdownRenderer;
import org.jellyfin.androidtv.util.Utils;
import org.jellyfin.apiclient.model.dto.BaseItemDto;

public class MyDetailsOverviewRowPresenter extends RowPresenter {
    private final MarkdownRenderer markdownRenderer;
    private ViewHolder viewHolder;

    public MyDetailsOverviewRowPresenter(MarkdownRenderer markdownRenderer) {
        super();

        this.markdownRenderer = markdownRenderer;

        // Don't call setActivated() on views
        setSyncActivatePolicy(SYNC_ACTIVATED_CUSTOM);
    }

    public final class ViewHolder extends RowPresenter.ViewHolder {
        private TextView mGenreRow;
        private LinearLayout mInfoRow;
        private TextView mTitle;
        private AsyncImageView mPoster;
        private TextView mSummary;
        private LinearLayout mButtonRow;
        private ProgressBar mProgress;

        private TextView mInfoTitle1;
        private TextView mInfoTitle2;
        private TextView mInfoTitle3;
        private TextView mInfoValue1;
        private TextView mInfoValue2;
        private TextView mInfoValue3;

        private RelativeLayout mLeftFrame;

        /**
         * Constructor for ViewHolder.
         *
         * @param rootView The View bound to the Row.
         * @param binding
         */
        public ViewHolder(DetailRowView view) {
            super(view);

            ViewRowDetailsBinding binding = view.getBinding();

            mTitle = binding.fdTitle;
            mInfoTitle1 = binding.infoTitle1;
            mInfoTitle2 = binding.infoTitle2;
            mInfoTitle3 = binding.infoTitle3;
            mInfoValue1 = binding.infoValue1;
            mInfoValue2 = binding.infoValue2;
            mInfoValue3 = binding.infoValue3;

            mLeftFrame = binding.leftFrame;

            mGenreRow = binding.fdGenreRow;
            mInfoRow = binding.fdMainInfoRow;
            mPoster = binding.mainImage;
            mProgress = binding.fdProgress;
            mButtonRow = binding.fdButtonRow;
            mSummary = binding.fdSummaryText;
        }

        public void collapseLeftFrame() {
            ViewGroup.LayoutParams params = mLeftFrame.getLayoutParams();
            params.width = Utils.convertDpToPixel(view.getContext(), 100);
        }
    }

    @Override
    protected ViewHolder createRowViewHolder(ViewGroup parent) {
        DetailRowView view = new DetailRowView(parent.getContext());
        viewHolder = new ViewHolder(view);

        return viewHolder;
    }

    @Override
    protected void onBindRowViewHolder(RowPresenter.ViewHolder holder, Object item) {
        super.onBindRowViewHolder(holder, item);

        MyDetailsOverviewRow row = (MyDetailsOverviewRow) item;
        ViewHolder vh = (ViewHolder) holder;

        setTitle(row.getItem().getName());
        InfoLayoutHelper.addInfoRow(holder.view.getContext(), row.getItem(), vh.mInfoRow, false, false);
        addGenres(vh.mGenreRow, row.getItem());
        setInfo1(row.getInfoItem1());
        setInfo2(row.getInfoItem2());
        setInfo3(row.getInfoItem3());

        String posterUrl = row.getImageDrawable();
        vh.mPoster.load(posterUrl, null, null, 1.0, 0);
        int progress = row.getProgress();
        if (progress > 0 && posterUrl != null) {
            vh.mProgress.setProgress(progress);
            vh.mProgress.setVisibility(View.VISIBLE);
        }

        String summaryRaw = row.getSummary();
        if (summaryRaw != null)
            vh.mSummary.setText(markdownRenderer.toMarkdownSpanned(summaryRaw));

        switch (row.getItem().getBaseItemType()) {
            case Person:
                RelativeLayout.LayoutParams params = (RelativeLayout.LayoutParams) vh.mSummary.getLayoutParams();
                params.topMargin = 10;
                params.height = Utils.convertDpToPixel(vh.view.getContext(), 185);
                vh.mSummary.setMaxLines(9);
                vh.mGenreRow.setVisibility(View.GONE);
                vh.mInfoRow.setVisibility(View.GONE);
                vh.collapseLeftFrame();

                break;
        }

        vh.mButtonRow.removeAllViews();
        for (TextUnderButton button : row.getActions()) {
            vh.mButtonRow.addView(button);
        }
    }

    private void addGenres(TextView textView, BaseItemDto item) {
        textView.setText(TextUtils.join(" / ", item.getGenres()));
    }

    public void setTitle(String text) {
        viewHolder.mTitle.setText(text);
        if (text.length() > 28) {
            // raise it up a bit
            ((RelativeLayout.LayoutParams) viewHolder.mTitle.getLayoutParams()).topMargin = Utils.convertDpToPixel(viewHolder.view.getContext(), 55);
        }
    }

    public void setInfo1(InfoItem info) {
        if (info == null) {
            viewHolder.mInfoTitle1.setText("");
            viewHolder.mInfoValue1.setText("");
        } else {
            viewHolder.mInfoTitle1.setText(info.getLabel());
            viewHolder.mInfoValue1.setText(info.getValue());
        }
    }

    public void setInfo2(InfoItem info) {
        if (info == null) {
            viewHolder.mInfoTitle2.setText("");
            viewHolder.mInfoValue2.setText("");
        } else {
            viewHolder.mInfoTitle2.setText(info.getLabel());
            viewHolder.mInfoValue2.setText(info.getValue());
        }
    }

    public void setInfo3(InfoItem info) {
        if (info == null) {
            viewHolder.mInfoTitle3.setText("");
            viewHolder.mInfoValue3.setText("");
        } else {
            viewHolder.mInfoTitle3.setText(info.getLabel());
            viewHolder.mInfoValue3.setText(info.getValue());
        }
    }

    public TextView getSummaryView() {
        return viewHolder.mSummary;
    }

    public void updateEndTime(String text) {
        if (viewHolder != null && viewHolder.mInfoValue3 != null)
            viewHolder.mInfoValue3.setText(text);
    }

    @Override
    protected void onSelectLevelChanged(RowPresenter.ViewHolder holder) {
        // Do nothing - this removes the shadow on the out of focus rows of image cards
    }
}

File: app/src/main/java/org/jellyfin/androidtv/ui/presentation/MyDetailsOverviewRowPresenter.kt
package org.jellyfin.androidtv.ui.presentation

import android.view.ViewGroup
import android.widget.RelativeLayout
import androidx.core.view.isVisible
import androidx.core.view.updateLayoutParams
import androidx.leanback.widget.RowPresenter
import org.jellyfin.androidtv.ui.DetailRowView
import org.jellyfin.androidtv.ui.itemdetail.MyDetailsOverviewRow
import org.jellyfin.androidtv.util.InfoLayoutHelper
import org.jellyfin.androidtv.util.MarkdownRenderer
import org.jellyfin.androidtv.util.dp
import org.jellyfin.sdk.model.api.BaseItemKind

class MyDetailsOverviewRowPresenter(
	private val markdownRenderer: MarkdownRenderer,
) : RowPresenter() {
	class ViewHolder(
		private val detailRowView: DetailRowView,
		private val markdownRenderer: MarkdownRenderer,
	) : RowPresenter.ViewHolder(detailRowView) {
		private val binding get() = detailRowView.binding

		fun setItem(row: MyDetailsOverviewRow) {
			setTitle(row.item.name)

			InfoLayoutHelper.addInfoRow(view.context, row.item, row.selectedMediaSourceIndex, binding.fdMainInfoRow, false, false)
			binding.fdGenreRow.text = row.item.genres?.joinToString(" / ")

			binding.infoTitle1.text = row.infoItem1?.label
			binding.infoValue1.text = row.infoItem1?.value

			binding.infoTitle2.text = row.infoItem2?.label
			binding.infoValue2.text = row.infoItem2?.value

			binding.infoTitle3.text = row.infoItem3?.label
			binding.infoValue3.text = row.infoItem3?.value

			binding.mainImage.load(row.imageDrawable, null, null, 1.0, 0)

			if (row.progress > 0 && row.imageDrawable != null) {
				binding.fdProgress.progress = row.progress
				binding.fdProgress.isVisible = true
			}

			setSummary(row.summary)

			if (row.item.type == BaseItemKind.PERSON) {
				binding.fdSummaryText.updateLayoutParams<RelativeLayout.LayoutParams> {
					topMargin = 10
					height = 185.dp(view.context)
				}

				binding.fdSummaryText.maxLines = 9
				binding.fdGenreRow.isVisible = false
				binding.leftFrame.updateLayoutParams<RelativeLayout.LayoutParams> {
					width = 100.dp(view.context)
				}
			}

			binding.fdButtonRow.removeAllViews()
			for (button in row.actions) {
				val parent = button.parent
				if (parent is ViewGroup) parent.removeView(button)

				binding.fdButtonRow.addView(button)
			}
		}

		fun setTitle(title: String?) {
			binding.fdTitle.text = title
			if (binding.fdTitle.text.length > 28) {
				binding.fdTitle.updateLayoutParams<RelativeLayout.LayoutParams> {
					topMargin = 55.dp(view.context)
				}
			}
		}

		fun setSummary(summary: String?) {
			binding.fdSummaryText.text = summary?.let { markdownRenderer.toMarkdownSpanned(it) }
		}

		fun setInfoValue3(text: String?) {
			binding.infoValue3.text = text
		}
	}

	var viewHolder: ViewHolder? = null
		private set

	init {
		syncActivatePolicy = SYNC_ACTIVATED_CUSTOM
	}

	override fun createRowViewHolder(parent: ViewGroup): ViewHolder {
		val view = DetailRowView(parent.context)
		viewHolder = ViewHolder(view, markdownRenderer)
		return viewHolder!!
	}

	override fun onBindRowViewHolder(viewHolder: RowPresenter.ViewHolder?, item: Any?) {
		super.onBindRowViewHolder(viewHolder, item)
		if (item !is MyDetailsOverviewRow) return
		if (viewHolder !is ViewHolder) return

		viewHolder.setItem(item)
	}

	override fun onSelectLevelChanged(holder: RowPresenter.ViewHolder) = Unit
}
